src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -260,6 +260,57 @@ public TypeSymbolWithAnnotations SetIsAnnotated(CSharpCompilation compilation)
         public TypeSymbolWithAnnotations AsNullableReferenceType() => _extensions.AsNullableReferenceType(this);
         public TypeSymbolWithAnnotations AsNotNullableReferenceType() => _extensions.AsNotNullableReferenceType(this);
 
+        public bool TryMergeNullability(TypeSymbolWithAnnotations other, VarianceKind variance, out TypeSymbolWithAnnotations merged)
+        {
+            TypeSymbol type;
+            bool? isNullable;
+            if (TypeSymbol.TryMergeNullability(other.TypeSymbol, out type) &&
+                TryMergeIsNullable(IsNullable, other.IsNullable, variance, out isNullable))
+            {
+                merged = Create(type, isNullable, CustomModifiers);
+                return true;
+            }
+            merged = default;
+            return false;
+        }
+
+        private static bool TryMergeIsNullable(bool? a, bool? b, VarianceKind variance, out bool? merged)
+        {
+            if (a == b)
+            {
+                merged = a;
+            }
+            else
+            {
+                switch (variance)
+                {
+                    case VarianceKind.In:
+                        merged = (a == false || b == false) ? (bool?)false : null;
+                        break;
+                    case VarianceKind.Out:
+                        merged = (a == true || b == true) ? (bool?)true : null;
+                        break;
+                    default:
+
+                        if (a == null)
+                        {
+                            merged = b;
+                        }
+                        else if (b == null)
+                        {
+                            merged = a;
+                        }
+                        else
+                        {
+                            merged = null;
+                            return false;
```

Comment:
>false [](start = 35, length = 5)

Is the return value about whether a warning is required?

Reply:
Indirectly, yes. The return value indicates whether the merge succeeded. If it failed, the caller (either MethodTypeInferrer or BestTypeInferrer) will fail inference and NullableWalker will report a warning.

---
In reply to: [223453527](https://github.com/dotnet/roslyn/pull/29727#discussion_r223453527) [](ancestors = 223453527)